File: internal/cdp/cdp.go
// Package cdp provides Chrome DevTools Protocol connectivity for Publisher Rocket.
//
// Publisher Rocket is an Electron app. This package launches it with
// --remote-debugging-port=9222, connects via raw WebSocket, and exposes
// Eval/Click helpers that mirror the TypeScript cdp.ts from the VS Code extension.
package cdp

import (
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

const (
	cdpHost        = "127.0.0.1"
	cdpPort        = 9222
	connectTimeout = 45 * time.Second
	pollInterval   = 500 * time.Millisecond
	rocketAppPath  = "/Applications/Publisher Rocket.app"
)

// ── Result types ─────────────────────────────────────────────────────────────

// StatusResult is returned by CheckStatus and surfaced to the frontend.
type StatusResult struct {
	Running    bool   `json:"running"`
	CDPEnabled bool   `json:"cdpEnabled"`
	PageID     string `json:"pageId"`
	Error      string `json:"error,omitempty"`
}

// LaunchResult is returned by EnsureRocket and surfaced to the frontend.
type LaunchResult struct {
	Success bool   `json:"success"`
	PageID  string `json:"pageId"`
	Error   string `json:"error,omitempty"`
}

// PageTarget mirrors the CDP /json response object.
type PageTarget struct {
	ID                   string `json:"id"`
	Title                string `json:"title"`
	WebSocketDebuggerURL string `json:"webSocketDebuggerUrl"`
}

// Session holds an open CDP WebSocket connection.
type Session struct {
	conn   net.Conn
	cmdID  int
	buf    []byte
}

// ── Binary helpers ────────────────────────────────────────────────────────────

// resolveMacBinary reads Info.plist to find the inner Mach-O binary name.
func resolveMacBinary(appPath string) (string, error) {
	plistPath := filepath.Join(appPath, "Contents", "Info.plist")
	data, err := os.ReadFile(plistPath)
	if err != nil {
		return "", fmt.Errorf("cannot read Info.plist: %w", err)
	}
	re := regexp.MustCompile(`<key>CFBundleExecutable</key>\s*<string>([^<]+)</string>`)
	m := re.FindSubmatch(data)
	if m == nil {
		return "", fmt.Errorf("CFBundleExecutable not found in Info.plist")
	}
	return filepath.Join(appPath, "Contents", "MacOS", string(m[1])), nil
}

// cleanEnv returns a copy of the current environment with Electron/VS Code
// vars scrubbed so the child process launches as a normal Electron app.
func cleanEnv() []string {
	var env []string
	for _, kv := range os.Environ() {
		key := strings.SplitN(kv, "=", 2)[0]
		switch key {
		case "ELECTRON_RUN_AS_NODE", "ELECTRON_NO_ATTACH_CONSOLE", "ELECTRON_NO_ASAR", "NODE_OPTIONS":
			continue
		}
		env = append(env, kv)
	}
	return env
}

// ── Launch ────────────────────────────────────────────────────────────────────

// LaunchRocket spawns Publisher Rocket with --remote-debugging-port=9222.
// Returns an error if the process exits immediately.
func LaunchRocket() error {
	exe, err := resolveMacBinary(rocketAppPath)
	if err != nil {
		return fmt.Errorf("cannot resolve Publisher Rocket binary: %w", err)
	}

	cmd := exec.Command(exe, fmt.Sprintf("--remote-debugging-port=%d", cdpPort))
	cmd.Env = cleanEnv()
	cmd.Stdout = io.Discard
	cmd.Stderr = io.Discard

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("failed to spawn Publisher Rocket: %w", err)
	}

	// Give the process 2s to either crash or start successfully.
	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()

	select {
	case err := <-done:
		return fmt.Errorf("Publisher Rocket exited immediately: %v", err)
	case <-time.After(2 * time.Second):
		// Still running — detach and return.
		return nil
	}
}

// QuitRocket quits Publisher Rocket via AppleScript.
func QuitRocket() error {
	out, err := exec.Command("osascript", "-e", `quit app "Publisher Rocket"`).CombinedOutput()
	if err != nil {
		return fmt.Errorf("failed to quit Publisher Rocket: %s (%w)", strings.TrimSpace(string(out)), err)
	}
	return nil
}

// ── CDP /json polling ─────────────────────────────────────────────────────────

// GetPageTarget queries http://127.0.0.1:9222/json and returns the Publisher
// Rocket page target, or an error if none is found.
func GetPageTarget() (*PageTarget, error) {
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Get(fmt.Sprintf("http://%s:%d/json", cdpHost, cdpPort))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var pages []PageTarget
	if err := json.NewDecoder(resp.Body).Decode(&pages); err != nil {
		return nil, err
	}

	for i, p := range pages {
		title := strings.ToLower(p.Title)
		if p.Title == "Publisher Rocket" || p.Title == "KDP Rocket" || strings.Contains(title, "rocket") {
			// Rewrite localhost → 127.0.0.1 to avoid Node 17+ IPv6 trap.
			pages[i].WebSocketDebuggerURL = strings.ReplaceAll(
				p.WebSocketDebuggerURL, "ws://localhost:", fmt.Sprintf("ws://%s:", cdpHost),
			)
			return &pages[i], nil
		}
	}
	return nil, fmt.Errorf("no Publisher Rocket page target found on port %d", cdpPort)
}

// IsPortOpen returns true if port 9222 is accepting TCP connections.
func IsPortOpen() bool {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", cdpHost, cdpPort), 2*time.Second)
	if err != nil {
		return false
	}
	conn.Close()
	return true
}

// PollForTarget polls /json until a Rocket target appears or the timeout expires.
func PollForTarget() (*PageTarget, error) {
	deadline := time.Now().Add(connectTimeout)
	for time.Now().Before(deadline) {
		if t, err := GetPageTarget(); err == nil {
			return t, nil
		}
		time.Sleep(pollInterval)
	}
	return nil, fmt.Errorf("CDP connection timed out after %s — Publisher Rocket did not expose a page target on port %d", connectTimeout, cdpPort)
}

// EnsureRocket handles all three cases:
//  1. Port not open → launch, poll.
//  2. Port open, no Rocket target → quit, relaunch, poll.
//  3. Port open, Rocket target found → return it.
func EnsureRocket() (*PageTarget, error) {
	target, err := GetPageTarget()
	if err == nil {
		return target, nil
	}

	if !IsPortOpen() {
		// Case 1: launch fresh.
		if err := LaunchRocket(); err != nil {
			return nil, err
		}
		time.Sleep(5 * time.Second)
		return PollForTarget()
	}

	// Case 2: port open but no Rocket target — quit and relaunch.
	_ = QuitRocket()
	time.Sleep(2 * time.Second)
	if err := LaunchRocket(); err != nil {
		return nil, err
	}
	time.Sleep(5 * time.Second)
	return PollForTarget()
}

// ── WebSocket session ─────────────────────────────────────────────────────────

// Connect opens a raw WebSocket connection to the CDP page endpoint.
func Connect(target *PageTarget) (*Session, error) {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", cdpHost, cdpPort), 5*time.Second)
	if err != nil {
		return nil, fmt.Errorf("CDP TCP connect failed: %w", err)
	}

	// WebSocket handshake.
	handshake := fmt.Sprintf(
		"GET /devtools/page/%s HTTP/1.1\r\nHost: %s:%d\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: YXV0aG9ydG9vbHMxMjM0PT0=\r\nSec-WebSocket-Version: 13\r\n\r\n",
		target.ID, cdpHost, cdpPort,
	)
	if _, err := conn.Write([]byte(handshake)); err != nil {
		conn.Close()
		return nil, fmt.Errorf("CDP handshake write failed: %w", err)
	}

	// Read past the HTTP 101 response.
	tmp := make([]byte, 4096)
	conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	if _, err := conn.Read(tmp); err != nil {
		conn.Close()
		return nil, fmt.Errorf("CDP handshake read failed: %w", err)
	}
	conn.SetReadDeadline(time.Time{})

	return &Session{conn: conn, cmdID: 1}, nil
}

// Close closes the CDP session.
func (s *Session) Close() {
	if s.conn != nil {
		s.conn.Close()
	}
}

// ── Send / Receive ────────────────────────────────────────────────────────────

// send writes a masked WebSocket text frame containing a CDP JSON message.
func (s *Session) send(method string, params map[string]interface{}) (int, error) {
	id := s.cmdID
	s.cmdID++

	payload, err := json.Marshal(map[string]interface{}{"id": id, "method": method, "params": params})
	if err != nil {
		return 0, err
	}

	mask := [4]byte{0xfe, 0xed, 0xbe, 0xef}
	masked := make([]byte, len(payload))
	for i, b := range payload {
		masked[i] = b ^ mask[i%4]
	}

	var header []byte
	if len(payload) <= 125 {
		header = []byte{0x81, byte(0x80 | len(payload))}
	} else {
		header = []byte{0x81, 0x80 | 126, byte(len(payload) >> 8), byte(len(payload))}
	}

	frame := append(header, mask[:]...)
	frame = append(frame, masked...)
	_, err = s.conn.Write(frame)
	return id, err
}

// recv reads WebSocket frames until it finds the response for targetID.
func (s *Session) recv(targetID int, timeout time.Duration) (interface{}, error) {
	s.conn.SetReadDeadline(time.Now().Add(timeout))
	defer s.conn.SetReadDeadline(time.Time{})

	for {
		header := make([]byte, 2)
		if _, err := io.ReadFull(s.conn, header); err != nil {
			return nil, fmt.Errorf("CDP recv header: %w", err)
		}

		isMasked := (header[1] & 0x80) != 0
		plen := int(header[1] & 0x7f)

		if plen == 126 {
			ext := make([]byte, 2)
			if _, err := io.ReadFull(s.conn, ext); err != nil {
				return nil, err
			}
			plen = int(ext[0])<<8 | int(ext[1])
		}

		if isMasked {
			if _, err := io.ReadFull(s.conn, make([]byte, 4)); err != nil {
				return nil, err
			}
		}

		data := make([]byte, plen)
		if _, err := io.ReadFull(s.conn, data); err != nil {
			return nil, err
		}

		var msg map[string]interface{}
		if err := json.Unmarshal(data, &msg); err != nil {
			continue
		}

		if id, ok := msg["id"].(float64); ok && int(id) == targetID {
			if result, ok := msg["result"].(map[string]interface{}); ok {
				if r, ok := result["result"].(map[string]interface{}); ok {
					return r["value"], nil
				}
			}
			return msg, nil
		}
	}
}

// Eval executes JavaScript in Publisher Rocket's renderer and returns the result as a string.
func (s *Session) Eval(expression string, timeout time.Duration) (string, error) {
	wrapped := fmt.Sprintf("(function(){ %s })()", expression)
	id, err := s.send("Runtime.evaluate", map[string]interface{}{
		"expression":    wrapped,
		"returnByValue": true,
	})
	if err != nil {
		return "", err
	}

	result, err := s.recv(id, timeout)
	if err != nil {
		return "", err
	}
	if result == nil {
		return "", nil
	}
	if s, ok := result.(string); ok {
		return s, nil
	}
	b, _ := json.Marshal(result)
	return string(b), nil
}

// Click dispatches a mouse press+release at (x, y) in the Rocket renderer.
func (s *Session) Click(x, y float64) error {
	params := map[string]interface{}{
		"type": "mousePressed", "x": x, "y": y,
		"button": "left", "clickCount": 1,
	}
	if _, err := s.send("Input.dispatchMouseEvent", params); err != nil {
		return err
	}
	time.Sleep(80 * time.Millisecond)

	params["type"] = "mouseReleased"
	if _, err := s.send("Input.dispatchMouseEvent", params); err != nil {
		return err
	}
	time.Sleep(80 * time.Millisecond)
	return nil
}

// KeyEscape dispatches an Escape keydown+keyup into the Rocket renderer.
func (s *Session) KeyEscape() {
	params := map[string]interface{}{"type": "keyDown", "key": "Escape", "code": "Escape"}
	s.send("Input.dispatchKeyEvent", params)
	time.Sleep(80 * time.Millisecond)
	params["type"] = "keyUp"
	s.send("Input.dispatchKeyEvent", params)
	time.Sleep(200 * time.Millisecond)
}

File: src-tauri/src/cdp.rs
// cdp.rs — Chrome DevTools Protocol connectivity for Publisher Rocket
//
// Launches Publisher Rocket with --remote-debugging-port=9222,
// connects via raw TCP WebSocket, exposes eval/click helpers.

use std::io::{Read, Write};
use std::net::TcpStream;
use std::os::unix::process::CommandExt;
use std::process::Command;
use std::time::{Duration, Instant};
use serde::Deserialize;
use serde_json::{json, Value};

const CDP_HOST: &str = "127.0.0.1";
const CDP_PORT: u16 = 9222;
const ROCKET_APP: &str = "/Applications/Publisher Rocket.app";

// ── Types ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize, Clone)]
pub struct PageTarget {
    pub id: String,
    pub title: String,
    #[serde(rename = "webSocketDebuggerUrl")]
    pub ws_url: String,
}

pub struct Session {
    stream: TcpStream,
    cmd_id: u32,
}

// ── Launch ────────────────────────────────────────────────────────────────────

/// Read CFBundleExecutable from Info.plist and return the full binary path.
fn resolve_mac_binary(app_path: &str) -> Result<String, String> {
    let plist = std::fs::read_to_string(format!("{}/Contents/Info.plist", app_path))
        .map_err(|e| format!("Cannot read Info.plist: {}", e))?;
    let re = regex::Regex::new(
        r"<key>CFBundleExecutable</key>\s*<string>([^<]+)</string>"
    ).unwrap();
    let exe = re.captures(&plist)
        .and_then(|c| c.get(1))
        .map(|m| m.as_str().to_string())
        .ok_or("CFBundleExecutable not found")?;
    Ok(format!("{}/Contents/MacOS/{}", app_path, exe))
}

/// Launch Publisher Rocket with --remote-debugging-port=9222.
pub fn launch_rocket() -> Result<(), String> {
    let exe = resolve_mac_binary(ROCKET_APP)?;

    let mut cmd = Command::new(&exe);
    cmd.arg(format!("--remote-debugging-port={}", CDP_PORT));

    // Scrub env vars that break Electron children
    for key in &["ELECTRON_RUN_AS_NODE", "ELECTRON_NO_ATTACH_CONSOLE",
                 "ELECTRON_NO_ASAR", "NODE_OPTIONS"] {
        cmd.env_remove(key);
    }

    // Detach — we don't own the process lifecycle
    unsafe { cmd.pre_exec(|| { libc::setsid(); Ok(()) }); }

    cmd.spawn().map_err(|e| format!("Failed to spawn Publisher Rocket: {}", e))?;

    // Give it 2s to either crash or stabilise
    std::thread::sleep(Duration::from_secs(2));
    Ok(())
}

/// Quit Publisher Rocket via AppleScript.
pub fn quit_rocket() -> Result<(), String> {
    Command::new("osascript")
        .args(["-e", r#"quit app "Publisher Rocket""#])
        .output()
        .map_err(|e| format!("Failed to quit Publisher Rocket: {}", e))?;
    Ok(())
}

// ── CDP /json polling ─────────────────────────────────────────────────────────

/// Query http://127.0.0.1:9222/json and return the Publisher Rocket target.
pub fn get_page_target() -> Result<PageTarget, String> {
    let resp = reqwest::blocking::Client::builder()
        .timeout(Duration::from_secs(3))
        .build()
        .unwrap()
        .get(format!("http://{}:{}/json", CDP_HOST, CDP_PORT))
        .send()
        .map_err(|e| e.to_string())?;

    let pages: Vec<PageTarget> = resp.json().map_err(|e| e.to_string())?;

    pages.into_iter()
        .find(|p| {
            p.title == "Publisher Rocket"
                || p.title == "KDP Rocket"
                || p.title.to_lowercase().contains("rocket")
        })
        .map(|mut t| {
            // Rewrite localhost → 127.0.0.1 to avoid IPv6 trap
            t.ws_url = t.ws_url.replace(
                "ws://localhost:",
                &format!("ws://{}:", CDP_HOST),
            );
            t
        })
        .ok_or_else(|| format!("No Publisher Rocket page target on port {}", CDP_PORT))
}

/// Check if port 9222 accepts TCP connections.
pub fn is_port_open() -> bool {
    TcpStream::connect_timeout(
        &format!("{}:{}", CDP_HOST, CDP_PORT).parse().unwrap(),
        Duration::from_secs(2),
    ).is_ok()
}

/// Poll /json until a Rocket target appears or 45s elapses.
pub fn poll_for_target() -> Result<PageTarget, String> {
    let deadline = Instant::now() + Duration::from_secs(45);
    while Instant::now() < deadline {
        if let Ok(t) = get_page_target() { return Ok(t); }
        std::thread::sleep(Duration::from_millis(500));
    }
    Err(format!(
        "CDP timeout — Publisher Rocket did not expose a page target on port {} within 45s",
        CDP_PORT
    ))
}

/// Ensure Rocket is running with CDP. Handles launch / relaunch as needed.
pub fn ensure_rocket() -> Result<PageTarget, String> {
    if let Ok(t) = get_page_target() { return Ok(t); }

    if !is_port_open() {
        // Case 1: nothing on port — fresh launch
        launch_rocket()?;
        std::thread::sleep(Duration::from_secs(5));
        return poll_for_target();
    }

    // Case 2: port open but no Rocket target — quit and relaunch
    let _ = quit_rocket();
    std::thread::sleep(Duration::from_secs(2));
    launch_rocket()?;
    std::thread::sleep(Duration::from_secs(5));
    poll_for_target()
}

// ── WebSocket session ─────────────────────────────────────────────────────────

/// Connect a raw WebSocket to the CDP page endpoint.
pub fn connect(target: &PageTarget) -> Result<Session, String> {
    let mut stream = TcpStream::connect_timeout(
        &format!("{}:{}", CDP_HOST, CDP_PORT).parse().unwrap(),
        Duration::from_secs(5),
    ).map_err(|e| format!("CDP TCP connect: {}", e))?;

    let handshake = format!(
        "GET /devtools/page/{} HTTP/1.1\r\n\
         Host: {}:{}\r\n\
         Upgrade: websocket\r\n\
         Connection: Upgrade\r\n\
         Sec-WebSocket-Key: YXV0aG9ydG9vbHMxMjM0PT0=\r\n\
         Sec-WebSocket-Version: 13\r\n\r\n",
        target.id, CDP_HOST, CDP_PORT
    );
    stream.write_all(handshake.as_bytes())
        .map_err(|e| format!("Handshake write: {}", e))?;

    // Read past HTTP 101
    stream.set_read_timeout(Some(Duration::from_secs(5))).ok();
    let mut buf = [0u8; 4096];
    stream.read(&mut buf).map_err(|e| format!("Handshake read: {}", e))?;
    stream.set_read_timeout(None).ok();

    Ok(Session { stream, cmd_id: 1 })
}

// ── Send / Recv ───────────────────────────────────────────────────────────────

impl Session {
    fn send(&mut self, method: &str, params: Value) -> Result<u32, String> {
        let id = self.cmd_id;
        self.cmd_id += 1;

        let payload = serde_json::to_vec(&json!({
            "id": id, "method": method, "params": params
        })).unwrap();

        let mask = [0xfe_u8, 0xed, 0xbe, 0xef];
        let masked: Vec<u8> = payload.iter().enumerate()
            .map(|(i, b)| b ^ mask[i % 4])
            .collect();

        let len = payload.len();
        let mut frame = Vec::new();
        frame.push(0x81);
        if len <= 125 {
            frame.push(0x80 | len as u8);
        } else {
            frame.push(0x80 | 126);
            frame.push((len >> 8) as u8);
            frame.push(len as u8);
        }
        frame.extend_from_slice(&mask);
        frame.extend_from_slice(&masked);

        self.stream.write_all(&frame)
            .map_err(|e| format!("CDP send: {}", e))?;
        Ok(id)
    }

    fn recv(&mut self, target_id: u32, timeout: Duration) -> Result<Value, String> {
        self.stream.set_read_timeout(Some(timeout)).ok();
        let result = self.recv_inner(target_id);
        self.stream.set_read_timeout(None).ok();
        result
    }

    fn recv_inner(&mut self, target_id: u32) -> Result<Value, String> {
        let mut buf = Vec::new();
        loop {
            let mut tmp = [0u8; 8192];
            let n = self.stream.read(&mut tmp)
                .map_err(|e| format!("CDP recv read: {}", e))?;
            buf.extend_from_slice(&tmp[..n]);

            // Try to parse frames from buffer
            let mut pos = 0;
            while pos + 2 <= buf.len() {
                let _fin_opcode = buf[pos]; pos += 1;
                let ml = buf[pos]; pos += 1;
                let is_masked = (ml & 0x80) != 0;
                let mut plen = (ml & 0x7f) as usize;

                if plen == 126 {
                    if pos + 2 > buf.len() { break; }
                    plen = ((buf[pos] as usize) << 8) | buf[pos + 1] as usize;
                    pos += 2;
                }
                if is_masked { pos += 4; }
                if pos + plen > buf.len() { break; }

                let frame = &buf[pos..pos + plen];
                pos += plen;

                if let Ok(msg) = serde_json::from_slice::<Value>(frame) {
                    if msg.get("id").and_then(|v| v.as_u64()) == Some(target_id as u64) {
                        let val = msg["result"]["result"]["value"].clone();
                        return Ok(if val.is_null() { msg } else { val });
                    }
                }
            }
        }
    }

    /// Evaluate JavaScript in Publisher Rocket's renderer.
    pub fn eval(&mut self, expression: &str, timeout_secs: u64) -> Result<String, String> {
        let wrapped = format!("(function(){{ {} }})()", expression);
        let id = self.send("Runtime.evaluate", json!({
            "expression": wrapped,
            "returnByValue": true
        }))?;
        let result = self.recv(id, Duration::from_secs(timeout_secs))?;
        Ok(match result {
            Value::String(s) => s,
            Value::Null => String::new(),
            other => other.to_string(),
        })
    }

    /// Dispatch a mouse press+release at (x, y).
    pub fn click(&mut self, x: f64, y: f64) -> Result<(), String> {
        self.send("Input.dispatchMouseEvent", json!({
            "type": "mousePressed", "x": x, "y": y,
            "button": "left", "clickCount": 1
        }))?;
        std::thread::sleep(Duration::from_millis(80));
        self.send("Input.dispatchMouseEvent", json!({
            "type": "mouseReleased", "x": x, "y": y,
            "button": "left", "clickCount": 1
        }))?;
        std::thread::sleep(Duration::from_millis(80));
        Ok(())
    }

    /// Dispatch Escape keydown+keyup.
    pub fn key_escape(&mut self) {
        let _ = self.send("Input.dispatchKeyEvent",
            json!({"type":"keyDown","key":"Escape","code":"Escape"}));
        std::thread::sleep(Duration::from_millis(80));
        let _ = self.send("Input.dispatchKeyEvent",
            json!({"type":"keyUp","key":"Escape","code":"Escape"}));
        std::thread::sleep(Duration::from_millis(200));
    }

    /// Type a single character using CDP Input.dispatchKeyEvent.
    /// Send only the 'char' event type — keyDown+char together causes double input.
    pub fn send_char(&mut self, ch: char) {
        let text = ch.to_string();
        let _ = self.send("Input.dispatchKeyEvent", json!({
            "type": "char",
            "key":  text,
            "text": text,
            "unmodifiedText": text,
        }));
        std::thread::sleep(Duration::from_millis(30));
    }

    /// Send Ctrl+A to select all text in the focused element.
    pub fn select_all(&mut self) {
        let _ = self.send("Input.dispatchKeyEvent", json!({
            "type": "keyDown", "key": "a",
            "modifiers": 2  // Ctrl
        }));
        std::thread::sleep(Duration::from_millis(30));
        let _ = self.send("Input.dispatchKeyEvent", json!({
            "type": "keyUp", "key": "a",
            "modifiers": 2
        }));
        std::thread::sleep(Duration::from_millis(30));
    }

    /// Send Backspace.
    pub fn send_backspace(&mut self) {
        let _ = self.send("Input.dispatchKeyEvent",
            json!({"type":"keyDown","key":"Backspace","code":"Backspace"}));
        std::thread::sleep(Duration::from_millis(20));
        let _ = self.send("Input.dispatchKeyEvent",
            json!({"type":"keyUp","key":"Backspace","code":"Backspace"}));
    }

    /// Tell Electron/Chrome to save downloads silently to a specific path.
    /// Must be called before clicking any export button.
    pub fn set_download_path(&mut self, path: &str) {
        // Browser.setDownloadBehavior works at browser context level
        let _ = self.send("Browser.setDownloadBehavior", json!({
            "behavior": "allow",
            "downloadPath": path,
            "eventsEnabled": true
        }));
        std::thread::sleep(Duration::from_millis(200));
        // Also set via Page (belt-and-suspenders)
        let _ = self.send("Page.setDownloadBehavior", json!({
            "behavior": "allow",
            "downloadPath": path
        }));
        std::thread::sleep(Duration::from_millis(200));
    }
}
